exec/internal/compile/scanner.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2019 The go-interpreter Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package compile
+
+import (
+	ops "github.com/go-interpreter/wagon/wasm/operators"
+)
+
+type scanner struct {
+	supportedOpcodes map[byte]bool
+}
+
+// InstructionMetadata describes a bytecode instruction.
+type InstructionMetadata struct {
+	Op    byte
+	Start int
+	Size  int
+}
+
+// CompilationCandidate describes a range of bytecode that can
+// be translated to native code.
+type CompilationCandidate struct {
+	// Bytecode index of the first opcode.
+	Beginning uint
+	// Bytecode index of the last byte in the instruction.
+	End uint
+	// InstructionMeta index of the first instruction.
+	StartInstruction int
+	// InstructionMeta index of the last instruction.
+	EndInstruction int
+	// Metrics about the instructions between first & last index.
+	Metrics Metrics
+}
+
+func (s *CompilationCandidate) reset() {
+	s.Beginning = 0
+	s.End = 0
+	s.StartInstruction = 0
+	s.EndInstruction = 0
+	s.Metrics = Metrics{}
+}
+
+// Bounds returns the beginning & end index in the bytecode which
+// this candidate would replace.
```

Comment:
we should probably document whether this is a half-open interval or not :)

Reply:
sorry, maths/physics-y speak, I guess :)
I meant to say we should probably say whether the interval we are describing holds the end part or excludes it.

_ie:_ 
- `[1, 2, 3, 4] --> [2, 3]` (when `Beginning=1` and `End=3`)
- `[1, 2, 3, 4] --> [2, 3, 4]` (when `Beginning=1` and `End=3`)

usual off-by-one errors... :)

Reply:
Ah, I see :)